tools/cldr-apps/src/org/unicode/cldr/unittest/web/TestMisc.java

```diff
@@ -82,27 +83,26 @@ public void TestIsCoverageOrganization() {
     }
 
     public void TestGitHash() {
-        final String appsVersion = CLDRConfigImpl.getGitHashForSlug("CLDR-Apps");
-        assertNotNull("getting CLDR-Apps version", appsVersion);
-        /*
-         * TODO: fail if CLDRURLS.UNKNOWN_REVISION.equals(appsVersion))
-         * and likewise for toolsVersion
-         */
-        // if (CLDRURLS.UNKNOWN_REVISION.equals(appsVersion)) {
-        //    errln("❌ appsVersion = UNKNOWN_REVISION: " + appsVersion);
-        // }
-
-        final String toolsVersion = CLDRConfigImpl.getGitHashForSlug("CLDR-Tools");
-        assertNotNull("getting CLDR-Tools version", toolsVersion);
-
-        /*
-         * TODO: add a regression test here for CLDR_DATA_HASH
-         * Reference: https://unicode-org.atlassian.net/browse/CLDR-13582
-         */
-        // final String hash = CLDRConfig.getInstance().getProperty("CLDR_DATA_HASH");
-        // assertNotNull("getting CLDR_DATA_HASH", hash);
-        // if (hash != null && !hash.matches("[0-9a-f]+")) {
-        //    errln("❌ CLDR_DATA_HASH is not hex: " + hash);
-        // }
+        final String hash = CldrUtility.getCldrBaseDirHash();
+        assertNotNull("getCldrBaseDirHash", hash);
+        if (hash != null) {
+            if (!hash.matches("[0-9a-f]+")) {
+                errln("❌ getCldrBaseDirHash is not hex: " + hash);
+            } else if (hash.length() < 8) {
+                errln("❌ getCldrBaseDirHash is shorter than 8 chars: " + hash);
+            }
+        }
+        final String strings[] = {"CLDR-Apps", "CLDR-Tools"};
+        for (String s : strings) {
+            final String version = CLDRConfigImpl.getGitHashForSlug(s);
+            assertNotNull("getting " + s + " version", version);
+            /*
+             * TODO: fail if CLDRURLS.UNKNOWN_REVISION.equals(version).
+             * Currently they do equal CLDRURLS.UNKNOWN_REVISION, since SurveyMain.init isn't called for TestAll.
```

Comment:
So, in a test environment, you definitely don't have `cldr-apps.war` and thus no manifest to read "CLDR-Apps" from. You also might not have `cldr*.jar` for its tools manifest, if the tests are run in eclipse let's say.  So this might be a warning, but a challenge to get from the test environment.

I don't know if you can use a warfile as if it were a jar on the classpath? If so, we could have a post build step, something like:   `java -DCLDR_APPS_VERSION_TEST -classpath …:…:cldr*.jar:cldr-apps.war TestAll` or something like that?

Reply:
I considered removing the getGitHashForSlug test entirely, or just the TODO. If you'd prefer either of those changes, that's OK with me.

That's all tangential to the cache-busting issue. I'm just concerned that a lot of code isn't under test at all, and the tight coupling with SurveyMain.init makes a lot of code hard to test. A unit test for getGitHashForSlug could be implemented using some other jar file, maybe a small one just for testing; or some other means of dependency injection. I don't know if you can use a warfile as if it were a jar on the classpath; it's very mysterious territory for me!